### PR TITLE
Add `StableLM`

### PR DESCRIPTION
**Pull request by jon-tow**
# What does this PR do?

This PR adds modeling support for [`StableLM 3B 4E1T`](https://huggingface.co/stabilityai/stablelm-3b-4e1t) (as well as [`StableLM 2 1.6B`](https://huggingface.co/stabilityai/stablelm-2-1_6b)) based models.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker

## Notes

**TODO**: The current online implementation uses an early naming scheme for the [`model_type`](https://huggingface.co/stabilityai/stablelm-3b-4e1t/blob/e3be657f4d1b78eb7520637ba922448a1ee456bd/config.json#L16)
```json
  "model_type": "stablelm_epoch",
```
I've temporarily created a development model repository https://huggingface.co/jon-tow/stablelm-3b-4e1t-dev for unit testing and [config archive mapping](https://github.com/jon-tow/transformers/blob/caf38d1659333ae826c0f64d2e3bec45e008081b/src/transformers/models/stablelm/configuration_stablelm.py#L24) which need to be updated before any merging.
Is there a better way to handle this? I've noticed a similar issue in [this](https://github.com/huggingface/transformers/pull/26170#discussion_r1368903469) `Phi` model PR discussion.

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Hey! Thanks for contributing! 
We usually recommend to start by adding the model on the hub, to allow for a quick distribution! 
See the tutorial [here](https://huggingface.co/docs/transformers/custom_models)

**Comment by jon-tow** [comment]
Hello!

> We usually recommend to start by adding the model on the hub, to allow for a quick distribution!

The model is already on the hub [here](https://huggingface.co/stabilityai/stablelm-3b-4e1t/tree/main) but uses custom modeling code. Is your suggestion to simply rename the `model_type` in the config.json and remove the custom implementation? Sorry if I'm misinterpreting this!

**Comment by ArthurZucker** [comment]
No what I mean is I think it's fine to keep it on the hub! 🤗 
We usually go for an integration if this is really asked by the community ( lots of activity on the repo / lots of activity on the issue for adding it here etc!) 
Thought it's really great that you want to contribute!
If you still want to add it, I would recommend you to make it as close as possible to other modelling files like Llama or Persimmon, and otherwise good that you created a repo for dev 👍🏻

File: src/transformers/models/stablelm/modeling_stablelm.py [hunk 1]
```diff
@@ -0,0 +1,1237 @@
...
+            query_layer,
+            key_layer,
+            value_layer,
+            indices_q,
+            (cu_seqlens_q, cu_seqlens_k),
+            (max_seqlen_in_batch_q, max_seqlen_in_batch_k),
+        )
+
+
+ATTENTION_CLASSES = {
```
Comment (ArthurZucker) [inline src/transformers/models/stablelm/modeling_stablelm.py:575]: I think it's worth adding the SDPA version, but can be done in a followup PR !

File: src/transformers/models/stablelm/modeling_stablelm.py [hunk 2]
```diff
@@ -0,0 +1,1237 @@
...
+    def set_decoder(self, decoder):
+        self.model = decoder
+
+    # Copied from transformers.models.llama.modeling_llama.LlamaForCausalLM.get_decoder
+    def get_decoder(self):
+        return self.model
+
+    @add_start_docstrings_to_model_forward(STABLELM_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=CausalLMOutputWithPast, config_class=_CONFIG_FOR_DOC)
+    def forward(
```
Comment (ArthurZucker) [inline src/transformers/models/stablelm/modeling_stablelm.py:963]:
```suggestion
    @replace_return_docstrings(output_type=CausalLMOutputWithPast, config_class=_CONFIG_FOR_DOC)
    # Ignore copy
    def forward(
```
adding copied from at the beginning of the class + this should work as well 😉

File: tests/models/stablelm/test_modeling_stablelm.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+            token_type_ids,
+            input_mask,
+            sequence_labels,
+            token_labels,
+            choice_labels,
+        ) = config_and_inputs
+        inputs_dict = {"input_ids": input_ids, "attention_mask": input_mask}
+        return config, inputs_dict
+
+
```
Comment (ArthurZucker) [inline tests/models/stablelm/test_modeling_stablelm.py:273]: Missing copied from here! (`with xxxxx->StableLm`)

File: tests/models/stablelm/test_modeling_stablelm.py [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+    import torch
+
+    from transformers import (
+        AutoTokenizer,
+        StableLMForCausalLM,
+        StableLMForSequenceClassification,
+        StableLMModel,
+    )
+
+
```
Comment (ArthurZucker) [inline tests/models/stablelm/test_modeling_stablelm.py:44]: copied from should be used here as well IMO! 🤗

**Review by ArthurZucker — COMMENTED** [review]
Looks already mergeable! Good work there 😉

File: tests/models/stablelm/test_modeling_stablelm.py (hunk 4 above)
Comment (jon-tow) [reply tests/models/stablelm/test_modeling_stablelm.py:44]: Good point 👍 I've added a copy-ignore comment to the constructor b/c this `num_key_value_heads` is a config field not used in Persimmon. Is this alright?

**Comment by jon-tow** [comment]
Hi, @ArthurZucker; thanks for the quick review! I'd like to point out that the recent commit https://github.com/huggingface/transformers/pull/28810/commits/097272f2272f545cf275c23e46cf7706e9bfac1f removes a copied-from comment from `StableLmModel` because `PersimmonModel` does not yet support `flash-attn` and the added `_attn_implementation` field breaks the `make repo-consistency` check. Let me know if you suggest a workaround 🙏

File: src/transformers/models/stablelm/modeling_stablelm.py [hunk 5]
```diff
@@ -0,0 +1,1240 @@
...
+    def set_output_embeddings(self, new_embeddings):
+        self.lm_head = new_embeddings
+
+    # Copied from transformers.models.llama.modeling_llama.LlamaForCausalLM.set_decoder
+    def set_decoder(self, decoder):
+        self.model = decoder
+
+    # Copied from transformers.models.llama.modeling_llama.LlamaForCausalLM.get_decoder
+    def get_decoder(self):
+        return self.model
```
Comment (ArthurZucker) [inline src/transformers/models/stablelm/modeling_stablelm.py:961]: ahh these are from persimmon I see haha alright!

File: src/transformers/models/stablelm/modeling_stablelm.py [hunk 6]
```diff
@@ -0,0 +1,1240 @@
...
+        'The weather is always wonderful in the San Juan Islands, and whether you're a vacationer or a resident, here are some ideas for fun!'
+        ```"""
+
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # decoder outputs consists of (dec_features, layer_state, dec_hidden, dec_attn)
```
Comment (ArthurZucker) [inline src/transformers/models/stablelm/modeling_stablelm.py:1011]:
```suggestion
```

File: tests/models/stablelm/test_modeling_stablelm.py (hunk 4 above)
Comment (ArthurZucker) [reply tests/models/stablelm/test_modeling_stablelm.py:46]: Of course!

File: tests/models/stablelm/test_modeling_stablelm.py [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+        StableLmForCausalLM,
+        StableLmForSequenceClassification,
+        StableLmModel,
+    )
+
+
+# Copied from transformers.tests.models.persimmon.test_modeling_persimmon.PersimmonModelTester with Persimmon -> StableLm
+class StableLmModelTester:
+    # Ignore copy
+    def __init__(
```
Comment (ArthurZucker) [inline tests/models/stablelm/test_modeling_stablelm.py:50]:
```suggestion

    # Ignore copy
    def __init__(
```

**Review by ArthurZucker — APPROVED** [review]
Great work there ! 
Looks good to me, would you just mind me merging #27931 before? 

This would mean you might have to use copied from mistral instead of Llama. Otherwise I'll merge this one and rebase on my side!

File: src/transformers/models/stablelm/configuration_stablelm.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+            Dictionary containing the scaling configuration for the RoPE embeddings. Currently supports two scaling
+            strategies: linear and dynamic. Their scaling factor must be a float greater than 1. The expected format is
+            `{"type": strategy name, "factor": scaling factor}`. When using this flag, don't update
+            `max_position_embeddings` to the expected new maximum. See the following thread for more information on how
+            these scaling strategies behave:
+            https://www.reddit.com/r/LocalLLaMA/comments/14mrgpr/dynamically_scaled_rope_further_increases/. This
+            is an experimental feature, subject to breaking API changes in future versions.
+        use_qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether or not the model should use bias for qkv layers.
+            Whether to tie weight embeddings
```
[the remark this answers is not in the document]
Comment (jon-tow) [reply src/transformers/models/stablelm/configuration_stablelm.py]: Great catch! Fixed 🙏

File: tests/models/stablelm/test_modeling_stablelm.py (hunk 7 above)
Comment (jon-tow) [reply tests/models/stablelm/test_modeling_stablelm.py:50]: Running `make style` reformats the new line between `class ...` and   `# Ignore copy` 😅

File: src/transformers/models/stablelm/configuration_stablelm.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+""" StableLM model configuration """
+
+from ...configuration_utils import PretrainedConfig
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+STABLELM_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "jon-tow/stablelm-3b-4e1t-dev": "https://huggingface.co/jon-tow/stablelm-3b-4e1t-dev/resolve/main/config.json",
```
[the remark this answers is not in the document]
Comment (jon-tow) [reply src/transformers/models/stablelm/configuration_stablelm.py]: Thanks for the reminder! I've opened draft PRs for the base models:
1. https://huggingface.co/stabilityai/stablelm-3b-4e1t/discussions/10
2. https://huggingface.co/stabilityai/stablelm-2-1_6b/discussions/6
At what point should these be merged? I assume after the next release of `transformers`?

File: src/transformers/models/stablelm/modeling_stablelm.py (hunk 1 above)
Comment (jon-tow) [reply src/transformers/models/stablelm/modeling_stablelm.py:575]: I can add it in a follow-up 🙏 I don't want to hold up the PR while I make sense of the changes needed 😅

File: src/transformers/models/stablelm/configuration_stablelm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/stablelm/configuration_stablelm.py]: Yes!

File: tests/models/stablelm/test_modeling_stablelm.py (hunk 7 above)
Comment (ArthurZucker) [reply tests/models/stablelm/test_modeling_stablelm.py:50]: Ah alright no worries then!

**Comment by ArthurZucker** [comment]
Could you rebase on main and make sure CIs are all green? 🤗 I can help if you can't finish all of them

**Comment by jon-tow** [comment]
Can you please help with the `test_hubs` workflow? It errors with `FAILED tests/trainer/test_trainer.py::TrainerIntegrationWithHubTester::test_push_to_hub_with_saves_each_epoch - AssertionError: 'Training in progress, epoch 1' not found in ['Training in progress, epoch 3', 'Training in progress, epoch 2', 'initial commit']` ([here](https://app.circleci.com/pipelines/github/huggingface/transformers/84263/workflows/f575df75-a763-4730-87af-5a533829762b/jobs/1088603?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary)). Not sure how to fix this one up 😅 Thanks!

**Review by ArthurZucker — APPROVED** [review]
Double checked, LGTM!

**Pull request merged**